mountpoint-s3/tests/fuse_tests/write_test.rs

```diff
@@ -807,6 +811,76 @@ where
     assert!(hello_contents.is_empty());
 }
 
+// This test checks that a write can be performed when IAM session policy inforces the usage of the specific SSE type and a KMS key ID
+// This test also contains error cases, that check that IAM session policy actually rejects writes with wrong SSE
+#[cfg(all(feature = "s3_tests", not(feature = "s3express_tests")))]
+#[test_case(ServerSideEncryption::new(None, None), true)]
+#[test_case(ServerSideEncryption::new(Some("aws:kms".to_owned()), None), true)]
+#[test_case(ServerSideEncryption::new(Some("aws:kms".to_owned()), Some(get_test_kms_key_id())), false)]
+fn write_with_sse_settings_test(sse: ServerSideEncryption, should_fail: bool) {
+    let sse_key = get_test_kms_key_id();
+
+    // configure credentials
+    let policy = r#"{"Statement": [
+        {"Effect": "Allow", "Action": ["s3:*"], "Resource": "*"},
+        {"Effect": "Allow", "Action": ["kms:*"], "Resource": "*"},
+        {
+            "Effect": "Deny",
+            "Action": ["s3:PutObject"],
+            "Resource": "*",
+            "Condition": {
+                "StringNotEquals": {
+                    "s3:x-amz-server-side-encryption": "aws:kms"
+                }
+            }
+        },
+        {
+            "Effect": "Deny",
+            "Action": ["s3:PutObject"],
+            "Resource": "*",
+            "Condition": {
+                "StringNotEquals": {
+                    "s3:x-amz-server-side-encryption-aws-kms-key-id": "__SSE_KEY_ARN__"
+                }
+            }
+        }
+    ]}"#;
+    let policy = policy.replace("__SSE_KEY_ARN__", &sse_key);
+
+    let mut test_config =
+        TestSessionConfig::default().with_credentials(tokio_block_on(get_scoped_down_credentials(&policy)));
+    test_config.filesystem_config.server_side_encryption = sse;
+    let (mount_point, _session, test_client) = fuse::s3_session::new("sse_with_policy_test", test_config);
+    let file_name = "hello";
+    let path = mount_point.path().join(file_name);
+    let mut f = open_for_write(&path, false, true).unwrap();
+    let data = vec![0xaa; 32];
+    let write_result = f.write_all(&data);
+
+    if should_fail {
+        assert!(
+            write_result.is_err(),
+            "should not be able to write to the file without proper sse"
+        );
+        return;
+    }
+
+    assert!(
+        write_result.is_ok(),
+        "should be able to write to the file with proper sse"
+    );
```

Comment:
prefer expect, it should hopefully provide better information to see why it wasn't OK?

Reply:
looks nicer for sure, will print an error on failure, thats true